tokio-timer/src/throttle.rs

```diff
@@ -0,0 +1,68 @@
+//! Slow down a stream by enforcing a delay between items.
+
+use {Delay, Error};
+
+use futures::{Async, Future, Poll, Stream};
+use futures::future::Either;
+
+use std::time::{Duration, Instant};
+
+/// Slow down a stream by enforcing a delay between items.
+#[derive(Debug)]
+pub struct Throttle<T> {
+    delay: Option<Delay>,
+    duration: Duration,
+    stream: Option<T>,
+}
+
+impl<T> Throttle<T> {
+    /// Slow down a stream by enforcing a delay between items.
+    pub fn new(stream: T, duration: Duration) -> Self {
+        Self {
+            delay: None,
+            duration: duration,
+            stream: Some(stream),
+        }
+    }
+}
+
+impl<T: Stream> Stream for Throttle<T> {
+    type Item = T::Item;
+    type Error = Either<T::Error, Error>;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        // 1. Is there a better way than nested match statements?
+        // 2. What about having a delay and the underlying stream error'd?
+
+        match self.delay.take() {
```

Comment:
I don't believe `take()` is necessary. You should be able to call poll directly.